File: examples/cpp/_6_FixedString/publisher.cpp
#include <iostream>
#include <thread>
#include "SharedPubSub.hpp"
#include "FixedString.h"
using namespace std;

int main(){
    
    shps::Publisher<FixedString<2048>> publisher("FixedString");
    FixedString<2048> value;
    
    while(1){
        /*--------- Example 1 : Publish ---------*/
        value = "EXAMPLE 1";
        publisher.publish(value);
        
        cout << "PUBLISHER : " << value.get() << " Normal publish"<< endl;
        this_thread::sleep_for(1s);

        /*--------- Example 2 : Publish only on change ---------*/
        
        for(int i=0;i<100;++i){
            publisher.publishOnChange(value);
        }
        value = "EXAMPLE 2";
        publisher.publishOnChange(value);

        cout << "PUBLISHER : " << value.get() << " Published on change" << endl;
        this_thread::sleep_for(1s);

        /*--------- Example 3 : Push multiple in queue and notify after ---------*/
        
        for(int i=0;i<10;++i){
            value = "EXAMPLE 3 "+to_string(i);
            publisher.setValueAndPush(value);
            cout << "PUBLISHER : " << value.get() << endl;
        }
        publisher.notifyAll();
        this_thread::sleep_for(1s);
    }
    return 0;
}

File: examples/nodejs/_5_Realtime/publisher.js
const SharedPubSub = require('../build/Release/SharedPubSub');

const publisher = new SharedPubSub.Publisher_int("Realtime");
let value = 0;

async function loop() {
    while (true) {
        // Publish the current value
        publisher.publish(value);
        console.log(`PUBLISHER : ${value} Normal publish`);
        value += 1; // Increment the value for the next publish
        await new Promise(resolve => setTimeout(resolve, 1000)); // Wait for 1 second

        // Example of publishing only on change
        if (value % 5 === 0) { // Publish only if the value is a multiple of 5
            publisher.publishOnChange(value);
            console.log(`PUBLISHER : ${value} Published on change`);
        }

        // Example of pushing multiple values in queue and notifying after
        for (let i = 0; i < 3; i++) {
            value += 1;
            publisher.setValueAndPush(value);
            console.log(`PUBLISHER : ${value}`);
        }
        publisher.notifyAll();
        console.log(`PUBLISHER : ${value} Sent Multiple`);
        await new Promise(resolve => setTimeout(resolve, 1000)); // Wait for 1 second
    }
}

loop().catch(console.error);